DAGE-25: main demo v1

nicolo-rinaldi opened the pull request:
Full pipeline in the file dataset_generator.py. Number of total queries fixed. Implemented the score of the full cartesian product between retrieved documents and queries.

[ticket](https://sease.atlassian.net/browse/DAGE-25)

nicolo-rinaldi committed:
[DAGE-15] changed config to have less value and started dataset_generator.py full pipeline

nicolo-rinaldi committed:
[DAGE-15] updated dataset_generator.py

nicolo-rinaldi committed:
[DAGE-15] full (provisory) pipeline coded in file dataset_generator.py

nicolo-rinaldi committed:
[DAGE-15] build of integration tests

Changes:
- added a test for utils.py functions
- started building the test_integration.py file to mock and test a full pipeline, to see if all the blocks interact with each other as expected

nicolo-rinaldi committed:
[DAGE-15] Added first version of main script and resolved the conflicts

nicolo-rinaldi committed:
[DAGE-25] First working demo

- Modified config.yaml to be launched with the demo
- updated dataset_generator.py to be a full pipeline. Right now, it has errors (needed upgrade in the data_store.py file)

nicolo-rinaldi committed:
[DAGE-25] First working demo

- Modified config.yaml to be launched with the demo
- updated dataset_generator.py to be a full pipeline. Right now, it has errors (needed upgrade in the data_store.py file)

nicolo-rinaldi committed:
[DAGE-25] minor changes to dataset_generator.py

nicolo-rinaldi committed:
[DAGE-25] Updated code according to data_store.py new features

nicolo-rinaldi committed:
[DAGE-25] added empty lines

nicolo-rinaldi committed:
[DAGE-25] Modified main file

- modified dataset_generator.py file according to PR
- added factory design patterns for both search engines and writers

nicolo-rinaldi committed:
[DAGE-25] Moved factory for llm and adapted code in dataset_generator.py

nicolo-rinaldi committed:
[DAGE-25] Added empty lines at the end of some files

nseidan commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,85 @@
+    log.debug(f"Number of documents evaluated: {len(docs_to_generate_queries)}")
+    for query_rating_context in data_store.get_queries():
+        for doc in data_store.get_documents():
+            data_store.add_query(query_rating_context.get_query(), doc.id)
```
adding new queries to the data_store while iterating over ` data_store.get_queries()` ?  -> you might have subtle bugs there ? it's better to store `data_store.get_queries() `in a local var then iterate over this.

nicolo-rinaldi replied:
fixed

nseidan commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,85 @@
+
+    user_queries = []
+    if config.queries is not None:
+        with open(config.queries, 'r', encoding='utf-8') as file:
```
is the file valid/already validated? no try/catch ?

nicolo-rinaldi replied:
Yes, if the file is not there, pydantic's `Config` model is going to throw an error.

nseidan commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,85 @@
+                flag = True
+                break
+            query_id = data_store.add_query(query_text, doc.id)
+            data_store.add_rating_score(query_id, doc.id, 2)
```
assuming score type is graded not binary ?  because i see 2.

nicolo-rinaldi replied:
yes, you are absolutely right. Fixed

nseidan commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,85 @@
+    llm = build_chat_model(LLMConfig.load(config.llm_configuration_file))
+    service = LLMService(chat_model=llm)
+
+    num_queries_per_doc = int(( (config.num_queries_needed - len(user_queries)) // config.doc_number) * 1.5)
```
btw if config .doc_number is 0, it will crash.

nicolo-rinaldi replied:
Pydantic's `Config` model is going to throw an error if it's not greater than 0

nicolo-rinaldi committed:
[DAGE-25] Changes requested in last PR
- added relevance scale label set to have it available in the Config variable
- modified dataset_generator.py loop through retrieved queries and assignment of max relevance label

dantuzi commented on rre-dataset-generator/src/search_engine/__init__.py:
```diff
@@ -0,0 +1,18 @@
+    "solr": SolrSearchEngine,
+}
+
+def build_search_engine(search_engine_type: str, endpoint: HttpUrl) -> BaseSearchEngine:
```
I'm expecting the factory as a dedicated class `SearchEngineFactory` with a static method `build(search_engine_type: str, endpoint: HttpUrl)`.

If I have to be completely honest, I don't like this method in the __init__.py file

nicolo-rinaldi replied:
refactored and changed the file of all 3 factories

dantuzi commented on rre-dataset-generator/src/llm/llm_service.py:
```diff
@@ -22,7 +23,7 @@ def generate_queries(self, document: Document, num_queries_generate_per_doc: int
         system_prompt = (
             f"You are a helpful assistant! Generate {num_queries_generate_per_doc} "
-            "keyword-phrase-based queries based on the given document below. "
+            "queries based on the given document below. " #keyword-phrase-based
```
I guess we can directly remove the comment after the prompt

nicolo-rinaldi replied:
removed

dantuzi commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,86 @@
         configure_logging(logging.INFO)
+    log = logging.getLogger(__name__)
+
+    search_engine = build_search_engine(search_engine_type=config.search_engine_type,
```
I was expecting something like
```
search_engine: BaseSearchEngine = SearchEngineFactory.build(
   search_engine_type = config.search_engine_type,
   endpoint=config.search_engine_collection_endpoint
)
```

or

```
search_engine: BaseSearchEngine = SearchEngineFactory.build(config=config)
```

nicolo-rinaldi replied:
done

dantuzi commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,86 @@
+    flag = False
+    for doc in docs_to_generate_queries:
+        data_store.add_document(doc.id, doc)
+        query_texts = service.generate_queries(doc, num_queries_per_doc)
```
the name `query_texts` can be replaced with a simpler `queries`

nicolo-rinaldi replied:
changed

dantuzi commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,86 @@
+    num_queries_per_doc = int(( (config.num_queries_needed - len(user_queries)) // config.doc_number) * 1.5)
+
+    # query generation step
+    flag = False
```
The name `flag` is not meaningful; I would have preferred something like `all_queries_generated`.

Alternatively, a probably cleaner solution is to write a specific function that returns when all queries have been generated

nicolo-rinaldi replied:
changed the name of the flag

dantuzi commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,86 @@
+        with open(config.queries, 'r', encoding='utf-8') as file:
+            for line in file:
+                if line.strip():
+                    user_queries.append(line.strip())
```
We are adding all queries provided by the user into the variable `user_queries`, and we never read from it.
I suggest directly adding all user queries to the datastore (we don't specify the docId, so it probably requires updating the datastore), so we've already have them ready when we calculate the cartesian product

dantuzi replied:
changed in the new version by @nicolo-rinaldi

dantuzi commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,27 +1,86 @@
+            break
+
+    log.debug(f"Number of documents evaluated: {len(docs_to_generate_queries)}")
+    queries_to_add = data_store.get_queries()
+    for query_rating_context in queries_to_add:
+        for doc in data_store.get_documents():
+            data_store.add_query(query_rating_context.get_query(), doc.id)
```
I don't understand why you need this

nicolo-rinaldi replied:
To do the cartesian product and match every query with every document

nicolo-rinaldi committed:
[DAGE-25] updated factories and other minor changes mentioned in the PR

nicolo-rinaldi committed:
[DAGE-25] new version of working pipeline at dataset_generator.py
- updated dataset_generator.py with a working pipeline
- updated data_store.py to support adding the queries without the documents id
- changed name of llm factory

nicolo-rinaldi committed:
[DAGE-25] added empty lines at the end of files
